python/foglamp/plugins/south/systeminfo/systeminfo.py

```diff
@@ -0,0 +1,320 @@
+            while True:
+                _bytes = max_possible * struct_size
+                names = array.array('B')
+                for i in range(0, _bytes):
+                    names.append(0)
+                outbytes = struct.unpack('iL', fcntl.ioctl(
+                    s.fileno(),
+                    0x8912,  # SIOCGIFCONF
+                    struct.pack('iL', _bytes, names.buffer_info()[0])
+                ))[0]
+                if outbytes == _bytes:
+                    max_possible *= 2
+                else:
+                    break
+            namestr = names.tostring()
+            ifaces = []
+            for i in range(0, outbytes, struct_size):
+                iface_name = bytes.decode(namestr[i:i + 16]).split('\0', 1)[0]
+                iface_addr = socket.inet_ntoa(namestr[i + 20:i + 24])
+                ifaces.append((iface_name, iface_addr))
+
+            return ifaces
+
+        def interface_transmission(dev, direction):
+            """Return the transmisson rate of a interface under linux
+            https://stackoverflow.com/a/41448191
+            """
+            path = "/sys/class/net/{}/statistics/{}_bytes".format(dev, direction)
+            f = open(path, "r")
+            bytes_collected = int(f.read())
+            f.close()
+            return bytes_collected
+
+        network_interfaces = all_interfaces()
+        network_traffic = []
+        network_calc = {}
+        for interface_name, interface_ip in network_interfaces:
+            network_calc[interface_name] = {}
+            network_calc[interface_name]["bytes_recd_before"] = interface_transmission(interface_name, "rx")
+            network_calc[interface_name]["bytes_sent_before"] = interface_transmission(interface_name, "tx")
+
+        timestep = int(handle['networkSnifferPeriod']['value'])  # seconds
+        time.sleep(timestep)
+
+        for interface_name, interface_ip in network_interfaces:
+            network_calc[interface_name]["bytes_recd_after"] = interface_transmission(interface_name, "rx")
+            network_calc[interface_name]["bytes_sent_after"] = interface_transmission(interface_name, "tx")
+
+        for interface_name, interface_ip in network_interfaces:
+            network_traffic.append({
+                interface_name: {
+                    "IP": interface_ip,
+                    "numberOfNetworkPacketsReceived": network_calc[interface_name]["bytes_recd_after"] -
+                                                      network_calc[interface_name]["bytes_recd_before"],
+                    "numberOfNetworkPacketsSent": network_calc[interface_name]["bytes_sent_after"] -
+                                                  network_calc[interface_name]["bytes_sent_before"],
+                }
+            })
+        return network_traffic
+
+    def get_system_info():
+        data = {}
+        hostname = str(subprocess.Popen('hostname', shell=True, stdout=subprocess.PIPE).stdout.readlines()[0],
```

Comment:
please refactor this. 
check out, err and decode in next lines

Reply:
out.decode("utf-8") not working.